File: clients/client-python/tests/unittests/test_function_catalog.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.

import unittest
from unittest.mock import patch, Mock

from gravitino.api.function.function_change import FunctionChange
from gravitino.api.function.function_type import FunctionType
from gravitino.api.rel.types.types import Types
from gravitino.client.relational_catalog import RelationalCatalog
from gravitino.dto.audit_dto import AuditDTO
from gravitino.dto.function.function_definition_dto import FunctionDefinitionDTO
from gravitino.dto.function.function_dto import FunctionDTO
from gravitino.dto.function.function_impl_dto import (
    SQLImplDTO,
)
from gravitino.dto.function.function_param_dto import FunctionParamDTO
from gravitino.dto.responses.drop_response import DropResponse
from gravitino.dto.responses.entity_list_response import EntityListResponse
from gravitino.dto.responses.function_list_response import FunctionListResponse
from gravitino.dto.responses.function_response import FunctionResponse
from gravitino.exceptions.base import (
    NoSuchSchemaException,
    NoSuchFunctionException,
    FunctionAlreadyExistsException,
)
from gravitino.name_identifier import NameIdentifier
from gravitino.namespace import Namespace
from gravitino.utils import HTTPClient, Response


class TestFunctionCatalog(unittest.TestCase):
    metalake_name = "test_metalake"
    catalog_name = "test_catalog"
    schema_name = "test_schema"
    catalog_namespace = Namespace.of(metalake_name)

    @classmethod
    def setUpClass(cls) -> None:
        cls.rest_client = HTTPClient("http://localhost:8090")
        cls.catalog = RelationalCatalog(
            catalog_namespace=cls.catalog_namespace,
            name=cls.catalog_name,
            catalog_type=RelationalCatalog.Type.RELATIONAL,
            provider="test_provider",
            audit=AuditDTO("anonymous"),
            rest_client=cls.rest_client,
        )

    def _get_mock_http_resp(self, json_str: str, return_code: int = 200):
        mock_http_resp = Mock()
        mock_http_resp.getcode.return_value = return_code
        mock_http_resp.read.return_value = json_str.encode("utf-8")
        mock_http_resp.info.return_value = None
        mock_http_resp.url = None
        mock_resp = Response(mock_http_resp)
        return mock_resp

    def _mock_function_dto(
        self, name, function_type, comment, deterministic
    ) -> FunctionDTO:
        params = [FunctionParamDTO(_name="param1", _data_type=Types.IntegerType.get())]
        impl = SQLImplDTO(
            _runtime="SPARK",
            _sql="SELECT param1 + 1",
            _resources=None,
            _properties={},
        )
        definition = FunctionDefinitionDTO(
            _parameters=params, _return_type=Types.IntegerType.get(), _impls=[impl]
        )
        return FunctionDTO(
            _name=name,
            _definitions=[definition],
            _function_type=function_type,
            _deterministic=deterministic,
            _comment=comment,
            _audit=AuditDTO(
                "creator", "2022-01-01T00:00:00Z", "modifier", "2022-01-01T00:00:00Z"
            ),
        )

    def test_list_functions(self):
        func1 = NameIdentifier.of(
            self.metalake_name, self.catalog_name, self.schema_name, "func1"
        )
        func2 = NameIdentifier.of(
            self.metalake_name, self.catalog_name, self.schema_name, "func2"
        )

        resp_body = EntityListResponse(_code=0, _idents=[func1, func2])
        mock_resp = self._get_mock_http_resp(resp_body.to_json())

        with patch(
            "gravitino.utils.http_client.HTTPClient.get", return_value=mock_resp
        ):
            functions = self.catalog.as_function_catalog().list_functions(
                Namespace.of(self.schema_name)
            )
            self.assertEqual(2, len(functions))
            self.assertEqual("func1", functions[0].name())
            self.assertEqual("func2", functions[1].name())

        # Test NoSuchSchemaException
        with patch(
            "gravitino.utils.http_client.HTTPClient.get",
            side_effect=NoSuchSchemaException("schema not found"),
        ):
            with self.assertRaises(NoSuchSchemaException):
                self.catalog.as_function_catalog().list_functions(
                    Namespace.of(self.schema_name)
                )

    def test_list_function_infos(self):
        func1 = self._mock_function_dto("func1", FunctionType.SCALAR, "comment1", True)
        func2 = self._mock_function_dto("func2", FunctionType.SCALAR, "comment2", False)

        resp_body = FunctionListResponse(_code=0, _functions=[func1, func2])
        mock_resp = self._get_mock_http_resp(resp_body.to_json())

        with patch(
            "gravitino.utils.http_client.HTTPClient.get", return_value=mock_resp
        ):
            functions = self.catalog.as_function_catalog().list_function_infos(
                Namespace.of(self.schema_name)
            )
            self.assertEqual(2, len(functions))
            self.assertEqual("func1", functions[0].name())
            self.assertEqual("func2", functions[1].name())
            self.assertEqual("comment1", functions[0].comment())
            self.assertEqual("comment2", functions[1].comment())

    def test_get_function(self):
        ident = NameIdentifier.of(self.schema_name, "func1")
        mock_function = self._mock_function_dto(
            "func1", FunctionType.SCALAR, "mock comment", True
        )
        resp_body = FunctionResponse(_code=0, _function=mock_function)
        mock_resp = self._get_mock_http_resp(resp_body.to_json())

        with patch(
            "gravitino.utils.http_client.HTTPClient.get", return_value=mock_resp
        ):
            func = self.catalog.as_function_catalog().get_function(ident)
            self.assertIsNotNone(func)
            self.assertEqual("func1", func.name())
            self.assertEqual("mock comment", func.comment())

        # Test NoSuchFunctionException
        with patch(
            "gravitino.utils.http_client.HTTPClient.get",
            side_effect=NoSuchFunctionException("function not found"),
        ):
            with self.assertRaises(NoSuchFunctionException):
                self.catalog.as_function_catalog().get_function(ident)

    def test_register_function(self):
        ident = NameIdentifier.of(self.schema_name, "func1")
        mock_function = self._mock_function_dto(
            "func1", FunctionType.SCALAR, "mock comment", True
        )
        resp_body = FunctionResponse(_code=0, _function=mock_function)
        mock_resp = self._get_mock_http_resp(resp_body.to_json())

        with patch(
            "gravitino.utils.http_client.HTTPClient.post", return_value=mock_resp
        ):
            func = self.catalog.as_function_catalog().register_function(
                ident,
                "mock comment",
                FunctionType.SCALAR,
                True,
                definitions=mock_function.definitions(),
            )
            self.assertIsNotNone(func)
            self.assertEqual("func1", func.name())
            self.assertEqual("mock comment", func.comment())

        # Test FunctionAlreadyExistsException
        with patch(
            "gravitino.utils.http_client.HTTPClient.post",
            side_effect=FunctionAlreadyExistsException("function already exists"),
        ):
            with self.assertRaises(FunctionAlreadyExistsException):
                self.catalog.as_function_catalog().register_function(
                    ident,
                    "mock comment",
                    FunctionType.SCALAR,
                    True,
                    definitions=mock_function.definitions(),
                )

    def test_alter_function(self):
        ident = NameIdentifier.of(self.schema_name, "func1")
        mock_function = self._mock_function_dto(
            "func1", FunctionType.SCALAR, "updated comment", True
        )
        resp_body = FunctionResponse(_code=0, _function=mock_function)
        mock_resp = self._get_mock_http_resp(resp_body.to_json())

        with patch(
            "gravitino.utils.http_client.HTTPClient.put", return_value=mock_resp
        ):
            func = self.catalog.as_function_catalog().alter_function(
                ident, FunctionChange.update_comment("updated comment")
            )
            self.assertIsNotNone(func)
            self.assertEqual("updated comment", func.comment())

        # Test NoSuchFunctionException
        with patch(
            "gravitino.utils.http_client.HTTPClient.put",
            side_effect=NoSuchFunctionException("function not found"),
        ):
            with self.assertRaises(NoSuchFunctionException):
                self.catalog.as_function_catalog().alter_function(
                    ident, FunctionChange.update_comment("updated comment")
                )

    def test_drop_function(self):
        ident = NameIdentifier.of(self.schema_name, "func1")
        resp_body = DropResponse(_code=0, _dropped=True)
        mock_resp = self._get_mock_http_resp(resp_body.to_json())

        with patch(
            "gravitino.utils.http_client.HTTPClient.delete", return_value=mock_resp
        ):
            dropped = self.catalog.as_function_catalog().drop_function(ident)
            self.assertTrue(dropped)

        # Test function not exists
        resp_body = DropResponse(_code=0, _dropped=False)
        mock_resp = self._get_mock_http_resp(resp_body.to_json())
        with patch(
            "gravitino.utils.http_client.HTTPClient.delete", return_value=mock_resp
        ):
            dropped = self.catalog.as_function_catalog().drop_function(ident)
            self.assertFalse(dropped)

File: clients/client-java/src/test/java/org/apache/gravitino/client/TestFunctionCatalog.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.apache.gravitino.client;

import static org.apache.hc.core5.http.HttpStatus.SC_CONFLICT;
import static org.apache.hc.core5.http.HttpStatus.SC_NOT_FOUND;
import static org.apache.hc.core5.http.HttpStatus.SC_OK;
import static org.apache.hc.core5.http.HttpStatus.SC_SERVER_ERROR;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.google.common.collect.ImmutableMap;
import java.time.Instant;
import org.apache.gravitino.Catalog;
import org.apache.gravitino.NameIdentifier;
import org.apache.gravitino.Namespace;
import org.apache.gravitino.dto.AuditDTO;
import org.apache.gravitino.dto.CatalogDTO;
import org.apache.gravitino.dto.function.FunctionDTO;
import org.apache.gravitino.dto.function.FunctionDefinitionDTO;
import org.apache.gravitino.dto.function.FunctionImplDTO;
import org.apache.gravitino.dto.function.FunctionParamDTO;
import org.apache.gravitino.dto.function.SQLImplDTO;
import org.apache.gravitino.dto.requests.CatalogCreateRequest;
import org.apache.gravitino.dto.responses.CatalogResponse;
import org.apache.gravitino.dto.responses.DropResponse;
import org.apache.gravitino.dto.responses.EntityListResponse;
import org.apache.gravitino.dto.responses.ErrorResponse;
import org.apache.gravitino.dto.responses.FunctionResponse;
import org.apache.gravitino.exceptions.FunctionAlreadyExistsException;
import org.apache.gravitino.exceptions.NoSuchFunctionException;
import org.apache.gravitino.exceptions.NoSuchSchemaException;
import org.apache.gravitino.function.Function;
import org.apache.gravitino.function.FunctionCatalog;
import org.apache.gravitino.function.FunctionDefinition;
import org.apache.gravitino.function.FunctionDefinitions;
import org.apache.gravitino.function.FunctionImpl;
import org.apache.gravitino.function.FunctionImpls;
import org.apache.gravitino.function.FunctionParam;
import org.apache.gravitino.function.FunctionType;
import org.apache.gravitino.rel.types.Type;
import org.apache.gravitino.rel.types.Types;
import org.apache.hc.core5.http.Method;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;

public class TestFunctionCatalog extends TestBase {

  protected static Catalog catalog;

  private static GravitinoMetalake metalake;

  protected static final String metalakeName = "testMetalake";

  protected static final String catalogName = "testCatalog";

  private static final String provider = "test";

  @BeforeAll
  public static void setUp() throws Exception {
    TestBase.setUp();

    metalake = TestGravitinoMetalake.createMetalake(client, metalakeName);

    CatalogDTO mockCatalog =
        CatalogDTO.builder()
            .withName(catalogName)
            .withType(CatalogDTO.Type.RELATIONAL)
            .withProvider(provider)
            .withComment("comment")
            .withProperties(ImmutableMap.of("k1", "k2"))
            .withAudit(
                AuditDTO.builder().withCreator("creator").withCreateTime(Instant.now()).build())
            .build();

    CatalogCreateRequest catalogCreateRequest =
        new CatalogCreateRequest(
            catalogName,
            CatalogDTO.Type.RELATIONAL,
            provider,
            "comment",
            ImmutableMap.of("k1", "k2"));
    CatalogResponse catalogResponse = new CatalogResponse(mockCatalog);
    buildMockResource(
        Method.POST,
        "/api/metalakes/" + metalakeName + "/catalogs",
        catalogCreateRequest,
        catalogResponse,
        SC_OK);

    catalog =
        metalake.createCatalog(
            catalogName,
            CatalogDTO.Type.RELATIONAL,
            provider,
            "comment",
            ImmutableMap.of("k1", "k2"));
  }

  @Test
  public void testListFunctions() throws JsonProcessingException {
    NameIdentifier func1 = NameIdentifier.of("schema1", "func1");
    NameIdentifier func2 = NameIdentifier.of("schema1", "func2");
    NameIdentifier expectedResultFunc1 =
        NameIdentifier.of(metalakeName, catalogName, "schema1", "func1");
    NameIdentifier expectedResultFunc2 =
        NameIdentifier.of(metalakeName, catalogName, "schema1", "func2");
    String functionPath =
        withSlash(formatFunctionRequestPath(Namespace.of(metalakeName, catalogName, "schema1")));

    EntityListResponse resp =
        new EntityListResponse(new NameIdentifier[] {expectedResultFunc1, expectedResultFunc2});
    buildMockResource(Method.GET, functionPath, null, resp, SC_OK);
    NameIdentifier[] functions = catalog.asFunctionCatalog().listFunctions(func1.namespace());

    Assertions.assertEquals(2, functions.length);
    Assertions.assertEquals(func1, functions[0]);
    Assertions.assertEquals(func2, functions[1]);

    // Throw schema not found exception
    ErrorResponse errResp =
        ErrorResponse.notFound(NoSuchSchemaException.class.getSimpleName(), "schema not found");
    buildMockResource(Method.GET, functionPath, null, errResp, SC_NOT_FOUND);
    Assertions.assertThrows(
        NoSuchSchemaException.class,
        () -> catalog.asFunctionCatalog().listFunctions(func1.namespace()),
        "schema not found");

    // Throw Runtime exception
    ErrorResponse errResp2 = ErrorResponse.internalError("internal error");
    buildMockResource(Method.GET, functionPath, null, errResp2, SC_SERVER_ERROR);
    Assertions.assertThrows(
        RuntimeException.class,
        () -> catalog.asFunctionCatalog().listFunctions(func1.namespace()),
        "internal error");
  }

  @Test
  public void testGetFunction() throws JsonProcessingException {
    NameIdentifier func = NameIdentifier.of("schema1", "func1");
    String functionPath =
        withSlash(
            formatFunctionRequestPath(Namespace.of(metalakeName, catalogName, "schema1"))
                + "/func1");

    FunctionDTO mockFunction =
        mockFunctionDTO(func.name(), FunctionType.SCALAR, "mock comment", true);
    FunctionResponse resp = new FunctionResponse(mockFunction);
    buildMockResource(Method.GET, functionPath, null, resp, SC_OK);
    Function loadedFunction = catalog.asFunctionCatalog().getFunction(func);
    Assertions.assertNotNull(loadedFunction);
    assertFunction(mockFunction, loadedFunction);

    // Throw schema not found exception
    ErrorResponse errResp =
        ErrorResponse.notFound(NoSuchSchemaException.class.getSimpleName(), "schema not found");
    buildMockResource(Method.GET, functionPath, null, errResp, SC_NOT_FOUND);
    Assertions.assertThrows(
        NoSuchSchemaException.class,
        () -> catalog.asFunctionCatalog().getFunction(func),
        "schema not found");

    // Throw function not found exception
    ErrorResponse errResp1 =
        ErrorResponse.notFound(NoSuchFunctionException.class.getSimpleName(), "function not found");
    buildMockResource(Method.GET, functionPath, null, errResp1, SC_NOT_FOUND);
    Assertions.assertThrows(
        NoSuchFunctionException.class,
        () -> catalog.asFunctionCatalog().getFunction(func),
        "function not found");
  }

  @Test
  public void testRegisterFunction() throws JsonProcessingException {
    NameIdentifier func = NameIdentifier.of("schema1", "func1");
    String functionPath =
        withSlash(formatFunctionRequestPath(Namespace.of(metalakeName, catalogName, "schema1")));

    FunctionDTO mockFunction =
        mockFunctionDTO(func.name(), FunctionType.SCALAR, "mock comment", true);

    FunctionResponse resp = new FunctionResponse(mockFunction);
    // Use null for request body to match any request body
    buildMockResource(Method.POST, functionPath, null, resp, SC_OK);

    Function registeredFunction =
        catalog
            .asFunctionCatalog()
            .registerFunction(
                func,
                "mock comment",
                FunctionType.SCALAR,
                true,
                FunctionDefinitions.of(mockDefinition(Types.StringType.get())));

    Assertions.assertNotNull(registeredFunction);
    assertFunction(mockFunction, registeredFunction);

    // Throw function already exists exception
    ErrorResponse errResp =
        ErrorResponse.alreadyExists(
            FunctionAlreadyExistsException.class.getSimpleName(), "function already exists");
    buildMockResource(Method.POST, functionPath, null, errResp, SC_CONFLICT);
    Assertions.assertThrows(
        FunctionAlreadyExistsException.class,
        () ->
            catalog
                .asFunctionCatalog()
                .registerFunction(
                    func,
                    "mock comment",
                    FunctionType.SCALAR,
                    true,
                    FunctionDefinitions.of(mockDefinition(Types.StringType.get()))),
        "function already exists");
  }

  @Test
  public void testAlterFunction() throws JsonProcessingException {
    NameIdentifier func = NameIdentifier.of("schema1", "func1");
    String functionPath =
        withSlash(
            formatFunctionRequestPath(Namespace.of(metalakeName, catalogName, "schema1"))
                + "/func1");

    FunctionDTO mockFunction =
        mockFunctionDTO(func.name(), FunctionType.SCALAR, "updated comment", true);

    FunctionResponse resp = new FunctionResponse(mockFunction);
    // Use null for request body to match any request body
    buildMockResource(Method.PUT, functionPath, null, resp, SC_OK);

    FunctionCatalog functionCatalog = catalog.asFunctionCatalog();
    Function alteredFunction =
        functionCatalog.alterFunction(
            func, org.apache.gravitino.function.FunctionChange.updateComment("updated comment"));

    Assertions.assertNotNull(alteredFunction);
    Assertions.assertEquals("updated comment", alteredFunction.comment());

    // Throw function not found exception
    ErrorResponse errResp =
        ErrorResponse.notFound(NoSuchFunctionException.class.getSimpleName(), "function not found");
    buildMockResource(Method.PUT, functionPath, null, errResp, SC_NOT_FOUND);
    Assertions.assertThrows(
        NoSuchFunctionException.class,
        () ->
            functionCatalog.alterFunction(
                func,
                org.apache.gravitino.function.FunctionChange.updateComment("updated comment")),
        "function not found");
  }

  @Test
  public void testDropFunction() throws JsonProcessingException {
    NameIdentifier func = NameIdentifier.of("schema1", "func1");
    String functionPath =
        withSlash(
            formatFunctionRequestPath(Namespace.of(metalakeName, catalogName, "schema1"))
                + "/func1");

    DropResponse resp = new DropResponse(true);
    buildMockResource(Method.DELETE, functionPath, null, resp, SC_OK);
    Assertions.assertTrue(catalog.asFunctionCatalog().dropFunction(func));

    // Return false when the function does not exist
    DropResponse resp1 = new DropResponse(false);
    buildMockResource(Method.DELETE, functionPath, null, resp1, SC_OK);
    Assertions.assertFalse(catalog.asFunctionCatalog().dropFunction(func));
  }

  private static FunctionDTO mockFunctionDTO(
      String name, FunctionType functionType, String comment, boolean deterministic) {
    FunctionDefinitionDTO definition = mockDefinitionDTO(Types.StringType.get());

    return FunctionDTO.builder()
        .withName(name)
        .withFunctionType(functionType)
        .withComment(comment)
        .withDeterministic(deterministic)
        .withDefinitions(new FunctionDefinitionDTO[] {definition})
        .withAudit(AuditDTO.builder().withCreator("creator").withCreateTime(Instant.now()).build())
        .build();
  }

  private static FunctionDefinitionDTO mockDefinitionDTO(Type returnType) {
    FunctionParamDTO[] params =
        new FunctionParamDTO[] {
          FunctionParamDTO.builder()
              .withName("param1")
              .withDataType(Types.IntegerType.get())
              .build()
        };

    SQLImplDTO impl =
        new SQLImplDTO(FunctionImpl.RuntimeType.SPARK.name(), null, null, "SELECT param1 + 1");

    return FunctionDefinitionDTO.builder()
        .withParameters(params)
        .withReturnType(returnType)
        .withImpls(new FunctionImplDTO[] {impl})
        .build();
  }

  private static FunctionDefinition mockDefinition(Type returnType) {
    FunctionParam[] params =
        new FunctionParam[] {
          FunctionParamDTO.builder()
              .withName("param1")
              .withDataType(Types.IntegerType.get())
              .build()
        };

    FunctionImpl[] impls =
        new FunctionImpl[] {
          FunctionImpls.ofSql(FunctionImpl.RuntimeType.SPARK, "SELECT param1 + 1")
        };

    return FunctionDefinitions.of(params, returnType, impls);
  }

  private static void assertFunction(FunctionDTO expected, Function actual) {
    Assertions.assertEquals(expected.name(), actual.name());
    Assertions.assertEquals(expected.functionType(), actual.functionType());
    Assertions.assertEquals(expected.comment(), actual.comment());
    Assertions.assertEquals(expected.deterministic(), actual.deterministic());
    Assertions.assertEquals(expected.definitions().length, actual.definitions().length);
    Assertions.assertEquals(
        expected.definitions()[0].returnType(), actual.definitions()[0].returnType());
  }

  private static String formatFunctionRequestPath(Namespace ns) {
    Namespace schemaNs = Namespace.of(ns.level(0), ns.level(1));
    return BaseSchemaCatalog.formatSchemaRequestPath(schemaNs) + "/" + ns.level(2) + "/functions";
  }
}
